Ensure we emit an allocator shim when only some crate types need one

bjorn3 opened the pull request:
Found this while trying to write a test for https://github.com/rust-lang/rust/pull/145955.

bjorn3 committed:
Ensure we emit an allocator shim when only some crate types need one

bjorn3 committed:
Fix typo in comment

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17300230287/job/49109197170) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49109197170) [(plain)](https://github.com/rust-lang/rust/commit/8bd5fb7e5a9371cc0be2b14e7de7185e41348d36/checks/49109197170/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/force-inlining/cross-crate.rs stdout ----

error: auxiliary build of /checkout/tests/ui/force-inlining/auxiliary/callees.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/force-inlining/auxiliary/callees.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/force-inlining/cross-crate/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type=lib" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/force-inlining/cross-crate/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustcVp6t40/list" "-Wl,--no-undefined-version" "/tmp/rustcVp6t40/symbols.o" "<1 object files omitted>" "/tmp/rustcVp6t40/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcVp6t40/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/force-inlining/cross-crate/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/force-inlining/cross-crate/auxiliary/libcallees.so" "-shared" "-Wl,-soname=libcallees.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/lto/debuginfo-lto.rs stdout ----

error: auxiliary build of /checkout/tests/ui/lto/auxiliary/debuginfo-lto-aux.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lto/auxiliary/debuginfo-lto-aux.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/debuginfo-lto/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-g" "--crate-type=rlib" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/debuginfo-lto/auxiliary"
stdout: none
--- stderr -------------------------------
warning: field `0` is never read
##[warning]  --> /checkout/tests/ui/lto/auxiliary/debuginfo-lto-aux.rs:3:35
   |
---
   = help: consider removing this field

error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustc6lNC65/list" "-Wl,--no-undefined-version" "/tmp/rustc6lNC65/symbols.o" "<1 object files omitted>" "/tmp/rustc6lNC65/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustc6lNC65/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/debuginfo-lto/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/debuginfo-lto/auxiliary/libdebuginfo_lto_aux.so" "-shared" "-Wl,-soname=libdebuginfo_lto_aux.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error; 2 warnings emitted
------------------------------------------


---- [ui] tests/ui/lto/dwarf-mixed-versions-lto.rs stdout ----

error: auxiliary build of /checkout/tests/ui/lto/auxiliary/dwarf-mixed-versions-lto-aux.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lto/auxiliary/dwarf-mixed-versions-lto-aux.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/dwarf-mixed-versions-lto/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-g" "--crate-type=rlib" "-Cdwarf-version=4" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/dwarf-mixed-versions-lto/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustcAEQSJl/list" "-Wl,--no-undefined-version" "/tmp/rustcAEQSJl/symbols.o" "<1 object files omitted>" "/tmp/rustcAEQSJl/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcAEQSJl/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/dwarf-mixed-versions-lto/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/dwarf-mixed-versions-lto/auxiliary/libdwarf_mixed_versions_lto_aux.so" "-shared" "-Wl,-soname=libdwarf_mixed_versions_lto_aux.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/lto/msvc-imp-present.rs stdout ----

error: auxiliary build of /checkout/tests/ui/lto/auxiliary/msvc-imp-present.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lto/auxiliary/msvc-imp-present.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/msvc-imp-present/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "thinlto" "-C" "codegen-units=8" "-C" "prefer-dynamic" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/msvc-imp-present/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustc4uueVz/list" "-Wl,--no-undefined-version" "/tmp/rustc4uueVz/symbols.o" "<2 object files omitted>" "/tmp/rustc4uueVz/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustc4uueVz/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/msvc-imp-present/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/msvc-imp-present/auxiliary/libmsvc_imp_present.so" "-shared" "-Wl,-soname=libmsvc_imp_present.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/sepcomp/sepcomp-lib-lto.rs stdout ----

error: auxiliary build of /checkout/tests/ui/sepcomp/auxiliary/sepcomp_lib.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/sepcomp/auxiliary/sepcomp_lib.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib-lto/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-C" "codegen-units=3" "--crate-type=rlib,dylib" "-g" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib-lto/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustcCGPPxe/list" "-Wl,--no-undefined-version" "/tmp/rustcCGPPxe/symbols.o" "<3 object files omitted>" "/tmp/rustcCGPPxe/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcCGPPxe/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib-lto/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib-lto/auxiliary/libsepcomp_lib.so" "-shared" "-Wl,-soname=libsepcomp_lib.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/sepcomp/sepcomp-lib.rs stdout ----

error: auxiliary build of /checkout/tests/ui/sepcomp/auxiliary/sepcomp_lib.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/sepcomp/auxiliary/sepcomp_lib.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-C" "codegen-units=3" "--crate-type=rlib,dylib" "-g" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustcjruR9U/list" "-Wl,--no-undefined-version" "/tmp/rustcjruR9U/symbols.o" "<3 object files omitted>" "/tmp/rustcjruR9U/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcjruR9U/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib/auxiliary/libsepcomp_lib.so" "-shared" "-Wl,-soname=libsepcomp_lib.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------
```

</details>

bjorn3 committed:
Correctly handle different crate types disagreeing if the allocator shim is exported

Previously it would attempt to export the allocator shim even linking
for a crate type which pulls in the allocator shim from a dylib rather
than locally defining it.

bors commented:
:umbrella: The latest upstream changes (presumably #145970) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

WaffleLapkin commented:
r? WaffleLapkin
@bors r+

bors commented:
:pushpin: Commit f4888c2a680aff0bd191704a8d44716eee94103e has been approved by `WaffleLapkin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=WaffleLapkin f4888c2a680aff0bd191704a8d44716eee94103e -->
<!-- homu: {"type":"Approved","sha":"f4888c2a680aff0bd191704a8d44716eee94103e","approver":"WaffleLapkin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit f4888c2a680aff0bd191704a8d44716eee94103e with merge c37dd0d33f8e4e0051fcdd09b44101a78fe1b137...
<!-- homu: {"type":"BuildStarted","head_sha":"f4888c2a680aff0bd191704a8d44716eee94103e","merge_sha":"c37dd0d33f8e4e0051fcdd09b44101a78fe1b137"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17361434060/job/49282070905) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49282070905) [(plain)](https://github.com/rust-lang/rust/commit/c37dd0d33f8e4e0051fcdd09b44101a78fe1b137/checks/49282070905/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
To only update this specific test, also pass `--test-args linking/mixed-allocator-shim.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/linking/mixed-allocator-shim.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=wasm32-wasip1" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/mixed-allocator-shim" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/wasm32-wasip1/native/rust-test-helpers" "--crate-type" "staticlib,dylib" "-Zstaticlib-prefer-dynamic"
stdout: none
--- stderr -------------------------------
warning: dropping unsupported crate type `dylib` for target `wasm32-wasip1`

warning: 1 warning emitted
```

</details>

bjorn3 committed:
Ignore test when dylibs are not supported

bjorn3 commented:
@bors try jobs=test-various
